### ISSUE TITLE
Record validation is too permissive

### ISSUE THREAD

**Issue by gth828r** [ChrisRx/quickavro#27]
In the record validation code at https://github.com/ChrisRx/quickavro/blob/2ed84315c5ff060e960686ab9dee175866d89919/src/convert.c#L392, the set of fields to validate within the record are taken from the schema.  The set of fields to validate should be taken from the record itself, otherwise the validation code is too permissive.

To illustrate this, pretend there are two record types which are both part of a union in an AVDL schema:
```
record A {
    int x;
}

Record B {
    int x;
    int y;
}

union {
    A,
    B
} U;
```

Now say I am serializing a record of type B, which by definition must have the y value present in the record.  If I try to serialize this, it will enter `python_to_union`, which will ultimately enter `validate` for the union.  The first record type it will try to validate against will be record A, since that is the first thing defined in the union.  Validate will check to see if my record has an `x` field in it and if it is of the correct type.  This check will pass, and thus the code will serialize the record as type A, dropping the value of `y`.

**Comment by gth828r** [comment]
I believe I have a fix.  I'll try to get that opened up for a pull request today.

**Comment by gth828r** [comment]
FYI, this same bug exists in the original Apache python implementation.  It doesn't appear to be an issue for Java or C++ Apache Avro bindings, because the in-memory objects explicitly require the user to set the union branch, which limits which schemas get validated against.  Since most python implementations of Avro do runtime union branch resolution by looking at fields in the record and trying to infer the union branch by recursive validation, they are susceptible to this same problem if they validate records by iterating over fields in the schema (instead of iterating over fields in the record).

**Comment by gth828r** [comment]
It turns out there is an open issue for this bug in the Apache implementation at https://issues.apache.org/jira/browse/AVRO-1343

There is some discussion about not fixing that specific issue, but the issue tracker also has a patch and it claims that the issue is targeting being closed in the 1.9.0 release.  Just including this info for context.

**Comment by ChrisRx** [comment]
Fixed in #28
